src/ApiHandler.ts

```diff
@@ -139,6 +125,69 @@ export default class ApiHandler {
 			}
 		}
 
+		const byteFee = api.consts.transactionPayment.transactionByteFee;
+		const extrinsicBaseWeight = api.consts.system.extrinsicBaseWeight;
+		const coefficients = api.consts.transactionPayment.weightToFee[0];
+		const coeffFrac = coefficients.coeff_frac;
+		const coeffInt = coefficients.coeff_integer;
+		
+		// TODO: generalize for terms that may have more degrees & may be negative
+		const degree = coefficients.degree;
+		const negative = coefficients.negative;
+
+		for (let idx = 0; idx < block.extrinsics.length; ++idx) {
+			const extrinsic = block.extrinsics[idx];
+			if (!extrinsic.isSigned) {
+				continue;
+			}
+
+			try {
+				if (api.runtimeVersion.specName.toString() === 'kusama') {
+					extrinsics[idx].info = await api.rpc.payment.queryInfo(extrinsic.toHex(), parentHash);
+				} else {
+					const parsedXt = extrinsics[idx];
+					const xtEvents = parsedXt.events;
+					const success = 'system.ExtrinsicSuccess';
+					const failure = 'system.ExtrinsicFailed';
+					const completedEvent = xtEvents.find((event) => event.method === success || event.method === failure);
+					if (!completedEvent) {
+						parsedXt.info = { error: 'Unable to fetch fee info' };
+						continue;
+					}
+
+					const completedData = completedEvent.data;
+					if (!completedData) {
+						parsedXt.info = { error: 'Unable to fetch fee info' };
+						continue;
+					}
+
+					const weightInfo = completedData[completedData.length - 1] as DispatchInfo;
```

Comment:
Your code is correct, but I would love to see a comment on this line, it's not straightforward imo.
In case of `ExtrinsicSuccess`, the data is `[DispatchInfo]`. In case of `ExtrinsicFailed`, the data is `[DispatchError, DispatchInfo]`. That's why we take the last element of the array.

https://github.com/polkadot-js/api/blob/e3abb9511169022f9feeac432b0397f1b377aa79/packages/metadata/src/Metadata/v11/static-substrate.json#L399-L418

Reply:
Added a comment in an attempt to clarify this.